src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+//! Python coroutine implementation, used notably when wrapping `async fn`
+//! with `#[pyfunction]`/`#[pymethods]`.
+use crate::exceptions::{PyRuntimeError, PyStopIteration};
+use crate::pyclass::IterNextOutput;
+use crate::sync::GILOnceCell;
+use crate::types::PyIterator;
+use crate::{intern, IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use futures::task::ArcWake;
+use pyo3_macros::{pyclass, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    #[allow(clippy::type_complexity)]
+    throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
+    waker: Option<Arc<PyFutureWaker>>,
+}
+
+impl Coroutine {
+    /// Wrap a boxed future in to a Python coroutine.
+    ///
+    /// Coroutine `send` polls the wrapped future, ignoring the value passed
+    /// (should always be `None` anyway).
+    ///
+    /// If `throw_callback` has been provided, coroutine `throw` calls it
+    /// with the exception passed, and poll the wrapped future.
+    /// Otherwise, the exception is reraised and the future dropped.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        future: Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>,
+        throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
+    ) -> Self {
+        Self {
+            future: Some(future),
+            throw_callback,
+            waker: None,
+        }
+    }
+
+    ///  Wrap a generic future into a Python coroutine.
+    pub fn from_future<F, T, E>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + Send + 'static,
+        T: IntoPy<PyObject> + Send,
+        E: Send,
+        PyErr: From<E>,
+    {
+        let wrap = async move {
+            let obj = future.await?;
+            // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
+            Ok(obj.into_py(unsafe { Python::assume_gil_acquired() }))
+        };
+        Self::new(Box::pin(wrap), None)
+    }
+
+    fn poll(
+        &mut self,
+        py: Python<'_>,
+        throw: Option<&PyAny>,
+    ) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        // raise if the coroutine has already been run to completion
+        let future_rs = match self.future {
+            Some(ref mut fut) => fut,
+            None => return Err(PyRuntimeError::new_err(COROUTINE_REUSED_ERROR)),
+        };
+        // pass thrown exception to throw callback if there is one, otherwise reraise it
+        if let Some(exc) = throw {
+            if let Some(cb) = &mut self.throw_callback {
+                cb(exc)
+            } else {
+                self.future.take();
```

Comment:
Please make dropping the future on the floor more explicit, e.g. using `let _ = ..` or `drop(..)` here and below.

Reply:
Do you prefer `self.future = None` instead?

Reply:
No, I think this is more implicit than `self.future.take()` because the point is not only modifying `self.future` but also why this mean for the value previously stored there.

Hence, I think `let _ = self.future.take();`  or `drop(self.future.take());` would be best. 

(These implicit drops have been shown to be one of the most common sources of bugs in unsafe code, i.e. forgetting to use `ptr::write` to suppress them for invalid values.)

Reply:
There is no `unsafe` code here, and I think the main point is not to drop the Rust future but to reset the field to `None`.
That's why I prefer in fact `self.future = None`.
But I will follow the maintainer directive here :)

Reply:
> There is no unsafe code here

The remark was only to illustrate that these implicit drops are a common source of confusion for Rust programmers. I know they are for me which why I asked for making them more explicit.

> I think the main point is not to drop the Rust future but to reset the field to None.

Maybe I do misunderstand then: My understanding was that this marks the coroutine as complete (by raising an exception in this case) _and_ cancels the underlying Rust future to keep Python and Rust in sync?

Admittedly, the situation is different in line 129 below where the Rust completes normally and we just want to avoid polling it again (i.e. `self.future = None;`) as the drop there does not really "cancel" anything, but I'd argue the uniformity of doing this in the same everywhere is preferable.

Reply:
I've reused `Coroutine::close` in `Coroutine::poll`, I find it way better now.